pkg/internal/source/kind.go

```diff
@@ -104,14 +131,61 @@ func (ks *Kind) String() string {
 // WaitForSync implements SyncingSource to allow controllers to wait with starting
 // workers until the cache is synced.
 func (ks *Kind) WaitForSync(ctx context.Context) error {
+	if func() bool {
+		ks.mu.Lock()
+		defer ks.mu.Unlock()
+
+		return ks.started == nil
+	}() {
+		return fmt.Errorf("cannot wait for sync on an unstarted source")
+	}
+
 	select {
 	case err := <-ks.started:
+		if err != nil {
+			// Try to stop the source to clean up any resources
+			// and collect any errors.
+			if stopErr := ks.Stop(); stopErr != nil {
+				err = kerrors.NewAggregate([]error{err, fmt.Errorf("failed to stop source: %w", stopErr)})
+			}
+		}
 		return err
 	case <-ctx.Done():
-		ks.startCancel()
 		if errors.Is(ctx.Err(), context.Canceled) {
 			return nil
 		}
 		return errors.New("timed out waiting for cache to be synced")
 	}
 }
+
+// Stop is internal and should be called only by the Controller to stop the Source. It should block until the
+// Source has stopped.
+func (ks *Kind) Stop() error {
+	// First check that the started channel is populated
+	// by holding the lock.
```

Comment:
nit: comment seems slightly off. We hold the lock while checking but we don't really check by holding? (I guess)